File: src/hexa_gait/hexa_gait/gait_node.py
"""Gait engine ROS node.

Subscribes to ``/gait/params`` (last-write-wins, no queue replay) and
publishes ``/legs/targets`` at 50 Hz. Builds an ``Engine`` + ``Tripod``
at init using the YAML in ``hexa_description`` (single source of truth
for body geometry and standing pose) and this package's ``config/gait.yaml``
(engine-internal knobs + cold-start fallbacks).

The node is intentionally thin: all gait logic lives in the pure-python
``Engine``. Tests live alongside the engine modules; this file owns
only the ROS plumbing.
"""

from __future__ import annotations

from pathlib import Path

import rclpy
import yaml
from ament_index_python.packages import get_package_share_directory
from geometry_msgs.msg import Point
from hexa_interfaces.msg import BodyPose as BodyPoseMsg
from hexa_interfaces.msg import GaitParams, LegState, LegTargets
from hexa_kinematics.leg_specs import load_leg_specs
from rclpy.node import Node
from std_msgs.msg import Empty, String

from .clock import LEG_NAMES
from .engine import (
    Engine,
    EngineConfig,
    build_leg_contexts,
    initial_stance_from_yaml,
    nominal_stance_from_yaml,
    reseat_geometry_from_yaml,
)
from .gaits.tripod import Tripod


PUBLISH_RATE_HZ = 50.0


def _load_engine_config(path: Path) -> EngineConfig:
    with path.open() as f:
        raw = yaml.safe_load(f)
    init_cfg = raw["initialize"]
    reseat_cfg = raw["reseat"]
    return EngineConfig(
        stride_length=float(raw["stride_length"]),
        min_cycle_time=float(raw["min_cycle_time"]),
        max_cycle_time=float(raw["max_cycle_time"]),
        duty_factor=float(raw["duty_factor"]),
        step_height=float(raw["step_height"]),
        swing_width=float(raw["swing_width"]),
        controller_dt=float(raw["controller_dt"]),
        recenter_swing_time=float(raw["recenter_swing_time"]),
        cmd_zero_tol=float(raw["cmd_zero_tol"]),
        forced_touchdown_delay=float(raw["forced_touchdown_delay"]),
        touchdown_settle_time=float(raw["touchdown_settle_time"]),
        init_pair_swing_time=float(init_cfg["pair_swing_time"]),
        init_lift_body_time=float(init_cfg["lift_body_time"]),
        init_swing_clearance=float(init_cfg["swing_clearance"]),
        init_place_feet_clearance=float(init_cfg["place_feet_clearance"]),
        reseat_settle_delay=float(reseat_cfg["settle_delay"]),
        reseat_height_change_threshold=float(reseat_cfg["height_change_threshold"]),
        reseat_pair_swing_time=float(reseat_cfg["pair_swing_time"]),
        reseat_swing_clearance=float(reseat_cfg["swing_clearance"]),
    )


def _load_coxa_to_bottom(geometry_path: Path) -> float:
    with geometry_path.open() as f:
        raw = yaml.safe_load(f)
    return float(raw["body"]["coxa_to_bottom"])


class GaitNode(Node):
    def __init__(self) -> None:
        super().__init__("gait_node")

        gait_share = Path(get_package_share_directory("hexa_gait")) / "config"
        desc_share = Path(get_package_share_directory("hexa_description")) / "config"

        self._cfg = _load_engine_config(gait_share / "gait.yaml")
        nominal = nominal_stance_from_yaml(
            desc_share / "geometry.yaml", desc_share / "standing_pose.yaml"
        )
        initial = initial_stance_from_yaml(desc_share / "geometry.yaml")
        coxa_to_bottom = _load_coxa_to_bottom(desc_share / "geometry.yaml")
        leg_contexts = build_leg_contexts(
            desc_share / "geometry.yaml", desc_share / "standing_pose.yaml"
        )
        leg_specs = load_leg_specs(desc_share / "geometry.yaml")
        reseat_geometry = reseat_geometry_from_yaml(
            desc_share / "geometry.yaml", desc_share / "standing_pose.yaml"
        )
        self._engine = Engine(
            config=self._cfg,
            strategy=Tripod(),
            nominal_stance=nominal,
            initial_stance=initial,
            coxa_to_bottom=coxa_to_bottom,
            leg_contexts=leg_contexts,
            leg_specs=leg_specs,
            reseat_geometry=reseat_geometry,
        )

        # Latest GaitParams. Walk-cycle knobs are no longer on the wire;
        # only the commanded velocity arrives via /gait/params.
        self._gait_name: str = "tripod"
        self._linear_x: float = 0.0
        self._linear_y: float = 0.0
        self._angular_z: float = 0.0

        self._sub_params = self.create_subscription(
            GaitParams, "/gait/params", self._on_params, 10
        )
        # Operator-gated FOLDED ↔ STAND toggle. The engine starts in
        # FOLDED and emits the initial_pose foot positions until this
        # message arrives. Once standing, a second press triggers the
        # symmetric warm-shutdown (STAND → FOLDING → FOLDED). The start
        # button on the joystick is the canonical source, but anything
        # publishing here counts. Stray presses outside FOLDED / STAND
        # are no-ops.
        self._sub_init = self.create_subscription(
            Empty, "/gait/initialize", self._on_init, 10
        )
        # Sniff /body/pose for the height axis only. The other fields
        # belong to the kinematics / posture chain, not the gait. The
        # engine watches the height for a settle window and runs its
        # reseat ladder when the user lets the chassis come to rest
        # at a new height — see Engine.set_target_height.
        self._sub_body_pose = self.create_subscription(
            BodyPoseMsg, "/body/pose", self._on_body_pose, 10
        )
        self._pub_targets = self.create_publisher(LegTargets, "/legs/targets", 10)
        # State broadcast for the posture chain: `hexa_posture` gates body
        # pose application on this so the user can't push the chassis
        # around while the legs are folded or mid-INITIALIZE / mid-FOLD.
        # Published every tick so late subscribers converge within one
        # 50 Hz interval — cheaper than configuring transient_local QoS
        # on both ends.
        self._pub_state = self.create_publisher(String, "/gait/state", 10)

        self._last_tick_ns: int | None = None
        self._timer = self.create_timer(1.0 / PUBLISH_RATE_HZ, self._tick)

        self.get_logger().info(
            f"gait_node up: strategy=tripod, stride_length={self._cfg.stride_length:.3f} m, "
            f"cycle_time in [{self._cfg.min_cycle_time:.2f}, {self._cfg.max_cycle_time:.2f}] s, "
            f"duty_factor={self._cfg.duty_factor:.2f}, step_height={self._cfg.step_height:.3f} m"
        )

    def _on_init(self, _msg: Empty) -> None:
        if self._engine.start_initialize():
            self.get_logger().info("start-button trigger received: FOLDED → INITIALIZE")
            return
        # Anywhere other than FOLDED: queue a fold request. The engine
        # consumes the flag the next time it sits in STAND at zero
        # height, so a press while RESEATING / STAND folds cleanly
        # once the legs are settled. A stray press during GAIT /
        # STOPPING is harmless — the request stays latched until the
        # engine returns to STAND.
        if self._engine.request_fold():
            self.get_logger().info(
                f"start-button trigger received: fold requested (engine in {self._engine.state.name})"
            )
            return
        self.get_logger().info(
            f"start-button trigger ignored: engine in state {self._engine.state.name}"
        )

    def _on_body_pose(self, msg: BodyPoseMsg) -> None:
        self._engine.set_target_height(float(msg.z))

    def _on_params(self, msg: GaitParams) -> None:
        if msg.gait_name and msg.gait_name != "tripod":
            # v1 only ships tripod; ignore but log so future gait
            # selection lands cleanly.
            self.get_logger().warn(
                f"GaitParams.gait_name={msg.gait_name!r} is unsupported; staying on tripod"
            )
        self._gait_name = "tripod"
        self._linear_x = float(msg.linear_x)
        self._linear_y = float(msg.linear_y)
        self._angular_z = float(msg.angular_z)

    def _tick(self) -> None:
        now_ns = self.get_clock().now().nanoseconds
        if self._last_tick_ns is None:
            dt = 1.0 / PUBLISH_RATE_HZ
        else:
            dt = (now_ns - self._last_tick_ns) * 1e-9
            if dt <= 0.0:
                # /clock can rewind during sim resets; skip this tick.
                self._last_tick_ns = now_ns
                return
        self._last_tick_ns = now_ns

        out = self._engine.update(
            dt=dt,
            v_body_xy=(self._linear_x, self._linear_y),
            omega_z=self._angular_z,
        )

        msg = LegTargets()
        msg.header.stamp = self.get_clock().now().to_msg()
        leg_states: list[LegState] = []
        for name in LEG_NAMES:
            leg = out[name]
            state = LegState()
            state.leg_name = name
            state.foot_target = Point(
                x=leg.foot_target[0],
                y=leg.foot_target[1],
                z=leg.foot_target[2],
            )
            state.phase = leg.phase
            state.stance = leg.stance
            leg_states.append(state)
        msg.legs = leg_states
        self._pub_targets.publish(msg)
        self._pub_state.publish(String(data=self._engine.state.value))


def main(args=None) -> None:
    rclpy.init(args=args)
    node = GaitNode()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/hexa_gait_cpp/src/gait_node.cpp
// Gait engine ROS node. Port of gait_node.py.
//
// Subscribes to /gait/params (last-write-wins), /gait/initialize, and
// /body/pose (height axis only); publishes /legs/targets and /gait/state at
// 50 Hz. Builds an Engine at init from hexa_description's YAML (single source of
// truth for body geometry and standing pose) and its own ROS parameters
// (engine-internal knobs; defaults mirror config/gait.yaml, which the launch
// files pass as the param source). All gait logic lives in the pure Engine;
// this file owns only the ROS plumbing.

#include <chrono>
#include <cstdint>
#include <memory>
#include <optional>
#include <string>
#include <vector>

#include <ament_index_cpp/get_package_share_directory.hpp>
#include <rclcpp/rclcpp.hpp>
#include <yaml-cpp/yaml.h>

#include <geometry_msgs/msg/point.hpp>
#include <hexa_interfaces/msg/body_pose.hpp>
#include <hexa_interfaces/msg/gait_params.hpp>
#include <hexa_interfaces/msg/leg_state.hpp>
#include <hexa_interfaces/msg/leg_targets.hpp>
#include <std_msgs/msg/empty.hpp>
#include <std_msgs/msg/string.hpp>

#include "hexa_gait_cpp/engine.hpp"
#include "hexa_gait_cpp/gaits/registry.hpp"
#include "hexa_gait_cpp/kinematics.hpp"
#include "hexa_gait_cpp/types.hpp"

namespace {

constexpr double kPublishRateHz = 50.0;

double load_coxa_to_bottom(const std::string& geometry_path) {
  const YAML::Node raw = YAML::LoadFile(geometry_path);
  return raw["body"]["coxa_to_bottom"].as<double>();
}

class GaitNode : public rclcpp::Node {
 public:
  GaitNode() : rclcpp::Node("gait_node") {
    const std::string desc_share =
        ament_index_cpp::get_package_share_directory("hexa_description") +
        "/config";
    const std::string geometry = desc_share + "/geometry.yaml";
    const std::string standing = desc_share + "/standing_pose.yaml";

    std::string default_gait;
    cfg_ = read_engine_config(default_gait);

    auto nominal = hexa_gait::nominal_stance_from_yaml(geometry, standing);
    auto initial = hexa_gait::initial_stance_from_yaml(geometry);
    const double coxa_to_bottom = load_coxa_to_bottom(geometry);
    auto leg_contexts = hexa_gait::build_leg_contexts(geometry, standing);
    auto leg_specs = hexa_gait::kin::load_leg_specs(geometry);
    auto reseat_geometry =
        hexa_gait::reseat_geometry_from_yaml(geometry, standing);

    engine_ = std::make_unique<hexa_gait::Engine>(
        cfg_, hexa_gait::strategies().at(default_gait)(), default_gait,
        std::move(nominal), std::move(initial), coxa_to_bottom,
        std::move(leg_contexts), std::make_optional(std::move(leg_specs)),
        std::make_optional(reseat_geometry));

    gait_name_ = default_gait;

    sub_params_ = create_subscription<hexa_interfaces::msg::GaitParams>(
        "/gait/params", 10,
        [this](hexa_interfaces::msg::GaitParams::SharedPtr msg) {
          on_params(*msg);
        });
    sub_init_ = create_subscription<std_msgs::msg::Empty>(
        "/gait/initialize", 10,
        [this](std_msgs::msg::Empty::SharedPtr msg) { on_init(*msg); });
    sub_body_pose_ = create_subscription<hexa_interfaces::msg::BodyPose>(
        "/body/pose", 10,
        [this](hexa_interfaces::msg::BodyPose::SharedPtr msg) {
          engine_->set_target_height(msg->z);
        });

    pub_targets_ =
        create_publisher<hexa_interfaces::msg::LegTargets>("/legs/targets", 10);
    pub_state_ = create_publisher<std_msgs::msg::String>("/gait/state", 10);

    // Wall timer at the publish rate; dt is computed from the ROS clock inside
    // tick() so the engine integrates in sim time when use_sim_time is set.
    timer_ = create_wall_timer(
        std::chrono::duration<double>(1.0 / kPublishRateHz),
        [this]() { tick(); });

    RCLCPP_INFO(get_logger(),
                "gait_node up: strategy=%s, stride_length=%.3f m, "
                "min_swing_time=%.2f s, max_swing_time=%.2f s, "
                "step_height=%.3f m",
                default_gait.c_str(), cfg_.stride_length, cfg_.min_swing_time,
                cfg_.max_swing_time, cfg_.step_height);
  }

 private:
  // Declare and read the engine knobs from this node's ros params. Defaults
  // mirror config/gait.yaml; the launch files pass that file (plus the "gait"
  // tuning-overlay block, layered last) as the param source, so a bare
  // `ros2 run` still boots on these defaults. The nested initialize:/reseat:
  // blocks in the YAML surface as dotted parameter names.
  hexa_gait::EngineConfig read_engine_config(std::string& default_gait_out) {
    hexa_gait::EngineConfig cfg;
    cfg.stride_length = declare_parameter<double>("stride_length", 0.1);
    cfg.min_swing_time = declare_parameter<double>("min_swing_time", 0.30);
    cfg.max_swing_time = declare_parameter<double>("max_swing_time", 0.4);
    cfg.step_height = declare_parameter<double>("step_height", 0.08);
    cfg.swing_width = declare_parameter<double>("swing_width", 0.0);
    cfg.controller_dt = declare_parameter<double>("controller_dt", 0.005);
    cfg.cmd_zero_tol = declare_parameter<double>("cmd_zero_tol", 1.0e-4);
    cfg.pause_debounce_delay =
        declare_parameter<double>("pause_debounce_delay", 0.4);
    cfg.pause_to_reseat_delay =
        declare_parameter<double>("pause_to_reseat_delay", 0.2);
    cfg.gait_change_pause_to_reseat_delay =
        declare_parameter<double>("gait_change_pause_to_reseat_delay", 0.1);
    cfg.max_reset_time = declare_parameter<double>("max_reset_time", 1.2);
    cfg.init_pair_swing_time =
        declare_parameter<double>("initialize.pair_swing_time", 0.4);
    cfg.init_lift_body_time =
        declare_parameter<double>("initialize.lift_body_time", 0.6);
    cfg.init_swing_clearance =
        declare_parameter<double>("initialize.swing_clearance", 0.04);
    cfg.init_place_feet_clearance =
        declare_parameter<double>("initialize.place_feet_clearance", 0.012);
    cfg.reseat_pose_settle_delay =
        declare_parameter<double>("reseat.pose_settle_delay", 0.5);
    cfg.reseat_height_change_threshold =
        declare_parameter<double>("reseat.height_change_threshold", 0.002);
    cfg.reseat_pair_swing_time =
        declare_parameter<double>("reseat.pair_swing_time", 0.2);
    cfg.reseat_pair_dwell_time =
        declare_parameter<double>("reseat.pair_dwell_time", 0.05);
    cfg.reseat_swing_clearance =
        declare_parameter<double>("reseat.swing_clearance", 0.025);

    default_gait_out = declare_parameter<std::string>("default_gait", "tripod");
    if (hexa_gait::strategies().find(default_gait_out) ==
        hexa_gait::strategies().end()) {
      throw std::runtime_error("default_gait=" + default_gait_out +
                               " not in STRATEGIES");
    }
    return cfg;
  }

  void on_init(const std_msgs::msg::Empty&) {
    if (engine_->start_initialize()) {
      RCLCPP_INFO(get_logger(),
                  "start-button trigger received: FOLDED -> INITIALIZE");
      return;
    }
    // Anywhere other than FOLDED: queue a fold request, consumed the next time
    // the engine sits in STAND at zero height.
    if (engine_->request_fold()) {
      RCLCPP_INFO(get_logger(),
                  "start-button trigger received: fold requested (engine in %s)",
                  hexa_gait::state_name(engine_->state()).c_str());
      return;
    }
    RCLCPP_INFO(get_logger(),
                "start-button trigger ignored: engine in state %s",
                hexa_gait::state_name(engine_->state()).c_str());
  }

  void on_params(const hexa_interfaces::msg::GaitParams& msg) {
    // Strategy switch arrives folded into GaitParams.gait_name. gait_name_
    // tracks the last *handled* request so a rejected swap is not retried on
    // every message; the engine's strategy_name stays the source of truth.
    if (!msg.gait_name.empty() && msg.gait_name != gait_name_) {
      if (engine_->set_strategy(msg.gait_name)) {
        if (engine_->pending_strategy_name().has_value() &&
            engine_->pending_strategy_name().value() == msg.gait_name) {
          RCLCPP_INFO(get_logger(),
                      "gait change to '%s' pending — pause-and-reseat sequence "
                      "running (engine in %s)",
                      msg.gait_name.c_str(),
                      hexa_gait::state_name(engine_->state()).c_str());
        } else {
          RCLCPP_INFO(get_logger(), "gait strategy switched to '%s'",
                      msg.gait_name.c_str());
        }
      } else {
        RCLCPP_WARN(get_logger(),
                    "gait change to '%s' dropped — gait locked during "
                    "engagement (engine in %s); request not retried",
                    msg.gait_name.c_str(),
                    hexa_gait::state_name(engine_->state()).c_str());
      }
      gait_name_ = msg.gait_name;
    }
    linear_x_ = msg.linear_x;
    linear_y_ = msg.linear_y;
    angular_z_ = msg.angular_z;
  }

  void tick() {
    const int64_t now_ns = get_clock()->now().nanoseconds();
    double dt;
    if (!last_tick_ns_.has_value()) {
      dt = 1.0 / kPublishRateHz;
    } else {
      dt = static_cast<double>(now_ns - last_tick_ns_.value()) * 1e-9;
      if (dt <= 0.0) {
        // /clock can rewind during sim resets; skip this tick.
        last_tick_ns_ = now_ns;
        return;
      }
    }
    last_tick_ns_ = now_ns;

    // DEBUG: trace engine state transitions. Remove once verified.
    const hexa_gait::EngineState prev_state = engine_->state();
    const auto out = engine_->update(dt, {linear_x_, linear_y_}, angular_z_);
    if (engine_->state() != prev_state) {
      RCLCPP_INFO(get_logger(), "[gait-debug] %s -> %s",
                  hexa_gait::state_name(prev_state).c_str(),
                  hexa_gait::state_name(engine_->state()).c_str());
    }

    hexa_interfaces::msg::LegTargets msg;
    msg.header.stamp = now();
    // msg.legs is a fixed-size std::array<LegState, 6>; fill by LEG_NAMES index.
    for (std::size_t i = 0; i < hexa_gait::LEG_NAMES.size(); ++i) {
      const std::string& name = hexa_gait::LEG_NAMES[i];
      const hexa_gait::LegOutput& leg = out.at(name);
      hexa_interfaces::msg::LegState state;
      state.leg_name = name;
      geometry_msgs::msg::Point point;
      point.x = leg.foot_target[0];
      point.y = leg.foot_target[1];
      point.z = leg.foot_target[2];
      state.foot_target = point;
      state.phase = leg.phase;
      state.stance = leg.stance;
      msg.legs[i] = state;
    }
    msg.master_phase = engine_->master_phase();
    pub_targets_->publish(msg);

    std_msgs::msg::String state_msg;
    state_msg.data = hexa_gait::state_value(engine_->state());
    pub_state_->publish(state_msg);
  }

  hexa_gait::EngineConfig cfg_;
  std::unique_ptr<hexa_gait::Engine> engine_;

  std::string gait_name_;
  double linear_x_ = 0.0;
  double linear_y_ = 0.0;
  double angular_z_ = 0.0;

  rclcpp::Subscription<hexa_interfaces::msg::GaitParams>::SharedPtr sub_params_;
  rclcpp::Subscription<std_msgs::msg::Empty>::SharedPtr sub_init_;
  rclcpp::Subscription<hexa_interfaces::msg::BodyPose>::SharedPtr sub_body_pose_;
  rclcpp::Publisher<hexa_interfaces::msg::LegTargets>::SharedPtr pub_targets_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_state_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::optional<int64_t> last_tick_ns_;
};

}  // namespace

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<GaitNode>());
  rclcpp::shutdown();
  return 0;
}
